modules/javafx.base/src/main/java/com/sun/javafx/binding/ExpressionHelper.java

```diff
@@ -188,86 +190,55 @@ protected void fireValueChangedEvent() {
 
     private static class Generic<T> extends ExpressionHelper<T> {
 
-        private InvalidationListener[] invalidationListeners;
-        private ChangeListener<? super T>[] changeListeners;
-        private int invalidationSize;
-        private int changeSize;
-        private boolean locked;
+        private Map<InvalidationListener, Integer> invalidationListeners = new LinkedHashMap<>();
+        private Map<ChangeListener<? super T>, Integer> changeListeners = new LinkedHashMap<>();
         private T currentValue;
+        private int weakChangeListenerGcCount = 2;
+        private int weakInvalidationListenerGcCount = 2;
```

Comment:
Why are these set to 2 and why do you need them at all? The previous implementation needed to grow and shrink the array so it had to keep these, but `Map` takes care of this for you.

Reply:
I agree, I kept these in as I wasn't sure if there was a need to manually force the garbage collection of weak listeners at the same rate as the original implementation.
Removing this would make sense to me also.

Reply:
As far as I know a LinkedHashMap doesn't automatically remove weak listener entries. The listener maps can be holding weak listeners as well as normal listeners.
So we need to keep the original behaviour from before where a count was checked on every addListener call and the weak references were purged from the array using the original calculation for this count.
Otherwise the map would never garbage collect these weak references.

The initial value of 2 for these counts was just the starting count although this is not necessarily strictly accurate. To be completely accurate then we would have to set the appropriate count in each constructor as follows:

i.e. in the Constructor with 2 InvalidationListeners:
weakChangeListenerGcCount = 0
weakInvalidationListenerGcCount = 2

in the Constructor with 2 ChangeListeners:
weakChangeListenerGcCount = 2
weakInvalidationListenerGcCount = 0

in the Constructor with 1 InvalidationListener and 1 ChangeListener:
weakChangeListenerGcCount = 1
weakInvalidationListenerGcCount = 1

Now, I could have used a WeakHashMap to store the listeners where it would automatically purge weak listeners but this doesn't maintain insertion order. Even though the specification doesn't mandate that listeners should be called back in the order they are registered, the unit tests failed when I didn't maintain order.

I am happy to remove this weak listener purge code (as it would make the code much simpler) but then we wouldn't automatically remove the weak listeners, but this may not be deemed a problem anyway?

Reply:
> So we need to keep the original behaviour from before where a count was checked on every addListener call and the weak references were purged from the array using the original calculation for this count.

The GC'd weak listeners do need to be purged, but why is the original behavior of the counters still applicable?

Reply:
Just wanted to keep a similar behaviour as before using the same calculation based originally on the size of the listeners array list and now based on the size of the map. So in theory the weak listeners should be trimmed at the same rate as before.
Happy to hear alternatives.

Reply:
Also bear in mind that the trimming of weak listeners is extremely slow as it has to iterate through each listener performing the weak listener test. We can't call this every time we add a listener as this would lock up the JavaFX thread completely (I tried it). 
I assume this is why the original calculation was used where it backs of the rate the weak listener trimming code was called as the array list grew.

Reply:
I honestly don't quite understand the old cleanup behavior of `(oldCapacity * 3)/2 + 1`. Why is it grown by x1.5? In your tests, can you try to change the cleanup threshold to higher and lower values and see what differences you get?

At the very least, the initial values of the counters should be set according to the specific constructor used.